### PR TITLE
feature(terraform): init terraform module

### PR DESCRIPTION
**Pull request by SollyzDev**
This PR creates a terraform module the creates the Forwarder lambda and allows the user to pass a list of log group names to create the needed Subscription filters.

### CODE REVIEW

**Comment by schehata** [comment]
hey @pecigonzalo regarding your comment previously:

```
Code wise/module wise, I dont think we should be doing the log-group subscriptions from Terraform. I think I mentioned the same in CF. Given we have Lambdas that can do the subscription backfill and dynamic addition of new log-groups, I don't think static resources (tf and cf) are suitable for handling dynamic resources like log-groups.
```

The problem with Subscriber/Unsubscriber module is that it created resources that Clouformation is not aware of. I wanted to avoid this pattern with terraform, and make it easier to find all the created subscription logs and easily maintain a list of log group names. I agree with you there should be a better way to handle the dynamic log groups. Wondering if current setup + listener only would fit, wdyt?

**Comment by pecigonzalo** [comment]
> The problem with Subscriber/Unsubscriber module is that it created resources that Clouformation is not aware of. I wanted to avoid this pattern with terraform, and make it easier to find all the created subscription logs and easily maintain a list of log group names. I agree with you there should be a better way to handle the dynamic log groups. Wondering if current setup + listener only would fit, wdyt?

I understand the reasoning. The challenge is that the dynamic nature of resource like cloudwatch log groups does not fit well with Terraform rather static and stateful nature. If AWS supported a single global "subscription config" for all CWLG (CloudWatch Log Group) I would certainly prefer to configure this via Terraform.

For example, as a "Cloud Engineer", as soon as a new CWLG is created I want to ensure its logs are propagated the logging pipeline.
If this is done via a static terraform config for which I have to populate names, I have to either preemptively create the group and subscription or do it after it was created and I notice my logs are missing.
Furthermore, this is a challenge because its likely that the code that provisions the resource that creates the CWLG and the place in which I configure this are different.

I think the idea of having a CWLG controller, that takes care of configuring them all (kind of what we did with the lambda, and something I've implemented myself at other places) is more maintainable, fit for porpouse and works well in most environments.
The way I generally configure this is with a single controller lambda that has multiple triggers:
- CWLG creation event -> Subscribe and set retention
- Periodic -> Check all CWLG and reconcile any drift
- Manual event -> Check a named CWLG and reconcile

So instead of having two ways of doing this, one that is a bit more subpar than the other (terraform), I would advocate for a single recommended approach. Particularly when anyone needing to configure this via Terraform only needs 1 or 2 resources at most, so its not something for which we add a ton of value in our module or they might get more value out of the settings of the module they might using already (be that a dedicated CWLG module or a Lambda or RDS module that already contains some CWLG settings).

File: .github/CODEOWNERS
Comment (pecigonzalo) [inline .github/CODEOWNERS:1]: Seems like this user is invalid

File: module/README.md
Comment (pecigonzalo) [inline module/README.md:5]:
```suggestion
## Axiom CloudWatch Forwarder Terraform Module
```

File: modules/subscriber/subscriber.tf
Comment (pecigonzalo) [inline modules/subscriber/subscriber.tf:20]: WDYT of interpolating the version here from a variable? That it would be easy to update the code to point to a new version when no Terraform changes are required.

File: modules/listener/listener.tf
Comment (pecigonzalo) [inline modules/listener/listener.tf:168]: This is implicit by `s3_bucket_name = aws_s3_bucket.cloudtrail[0].id` and I believe it can be removed

File: modules/listener/listener.tf
Comment (pecigonzalo) [inline modules/listener/listener.tf:55]: I think this does not depend on `enable_cloudtrail`. The listener always triggers on some events by Event Bridge right?

File: modules/listener/variables.tf
Comment (pecigonzalo) [inline modules/listener/variables.tf:18]: Im a bit confused by this one, is the goal to enable CloudTrail events to be forwarded to Axiom or something else?

File: modules/listener/listener.tf
Comment (SollyzDev) [reply modules/listener/listener.tf:55]: This part confuses me, Event Bridge depends on Cloudtrail to trigger those events, right? When I tested without Cloudtrail the triggering didn't happen.

File: modules/listener/variables.tf
Comment (SollyzDev) [reply modules/listener/variables.tf:18]: The goal is to trigger the *Listener* when new log groups are created, for some reason it doesn't get triggered without Cloudtrail being enabled.

File: .github/workflows/ci.yaml
Comment (SollyzDev) [reply .github/workflows/ci.yaml]: done

File: modules/listener/listener.tf
Comment (pecigonzalo) [reply modules/listener/listener.tf:55]: Yes, this permission must be created, but it should not be dependant on this flag, as without this permission it will not work.

File: modules/subscriber/subscriber.tf
Comment (SollyzDev) [reply modules/subscriber/subscriber.tf]: updated

File: modules/listener/listener.tf
Comment (pecigonzalo) [inline modules/listener/listener.tf:165]: I think this has to be removed from here and from Cloudformation. We should not enable a CloudTrail Trail. CloudTrail itself and cloudtrail events are always enabled as far as I know. They don't require a Trail to be enabled, and if they do, its something users should/would manage globally, not as part of a module like this IMO.

File: modules/subscriber/subscriber.tf
Comment (SollyzDev) [reply modules/subscriber/subscriber.tf:19]: on it.

File: modules/listener/variables.tf
Comment (pecigonzalo) [reply modules/listener/variables.tf:18]: That seems strange as I mention in https://github.com/axiomhq/axiom-cloudwatch-forwarder/pull/83#discussion_r1790000187.

How did we test this? The playground account should have gotten a trail anyway because we create those at the Org level.

File: modules/subscriber/subscriber.tf
Comment (SollyzDev) [reply modules/subscriber/subscriber.tf]: What I added is just replacing the version with:

```s3_key        = "axiom-cloudwatch-forwarder/v${var.forwarder_version}/forwarder.zip"```

Wondering if its better to replace the whole s3_key with a variable.

File: modules/subscriber/subscriber.tf
Comment (SollyzDev) [reply modules/subscriber/subscriber.tf]: updated with `forwarder_bucket` variable.

File: .github/workflows/ci.yaml
Comment (SollyzDev) [reply .github/workflows/ci.yaml:3]: tough to say, I was using it through my development in terraform module. So it might be useful when developing locally, I think we can assume for now that we will develop python code separately and then push it, and later update terraform.

Comment (SollyzDev) [reply .github/workflows/ci.yaml:3]: should we have separate versioning for the python code and terraform module?

Comment (pecigonzalo) [reply .github/workflows/ci.yaml:3]: Im primarily thinking we should protect this from pushing on a public PR, either here, through its creds, or through repo configuration, but we want to prevent a random user from using the credentials.

Comment (pecigonzalo) [reply .github/workflows/ci.yaml:3]: > should we have separate versioning for the python code and terraform module?

Possibly, but might complicate the release workflow. I think we can keep them the same for now.

File: modules/listener/variables.tf
Comment (SollyzDev) [reply modules/listener/variables.tf:18]: tested this in our dev account. I saw there was a trail already, but it didn't work. hmmm, now that I think of it, perhaps it was missing permissions, I will test again.

File: modules/listener/listener.tf
Comment (SollyzDev) [reply modules/listener/listener.tf:164]: > CloudTrail itself and cloudtrail events are always enabled as far as I know.

So Users don't have to enable this at all? Should we mention something about it in the docs?

Comment (pecigonzalo) [reply modules/listener/listener.tf:164]: I don't think they have to enable it at all. I only doubt because of what you mentioned during the test, but I would like to replicate because I don't think a CloudTrail Trail is required for CloudTrail based events to work.

Comment (SollyzDev) [reply modules/listener/listener.tf]: I have tested again on our dev account without creating a new trail and it worked. Can we test somewhere on a new account or something where there is no trails at all?

Comment (pecigonzalo) [reply modules/listener/listener.tf]: I don't think we can disable it on a particular account, as its enabled for the entire org. @a-khaledf do we have any test accounts outside the org?

Comment (SollyzDev) [reply modules/listener/listener.tf]: I removed the *Cloudtrail* resources from terraform module and Cloudformation. If you are confident with it we can move on.

**Comment by pecigonzalo** [comment]
@SollyzDev Added the missing suggestion (inline the policy) https://github.com/axiomhq/axiom-cloudwatch-forwarder/pull/88 along some minor style fixes. Feel free to merge, otherwise LGTM.

**Comment by pecigonzalo** [comment]
@SollyzDev `inline_policy` in the role is [deprecated](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_role#example-of-exclusive-inline-policies), I would suggest reverting those 2 commits and either merging the linked PR or doing it that way.

File: .github/workflows/ci.yaml
Comment (a-khaledf) [reply .github/workflows/ci.yaml:3]: This shouldn't be how we test usually, pushing to an s3 bucket on dev should be something done locally to test, not push to test.

Comment (a-khaledf) [reply .github/workflows/ci.yaml:3]: Please make ci.yml to validate lint and format terraform and run test cases if there is any but not to push to s3 bucket that you use for testing. it is a public PR, noone contributing would have access to this internal s3 bucket if he wants to contribute
we shouldn't block the PR though, let's talk about testing in a follow up PR please

Comment (SollyzDev) [reply .github/workflows/ci.yaml:3]: sounds good, we can move on with this one, and create a new PR for the CI.

**Comment by SollyzDev** [comment]
> @SollyzDev Added the missing suggestion (inline the policy) #88 along some minor style fixes. Feel free to merge, otherwise LGTM.

I enabled auto-merge, waiting for reviewers approval.

File: .github/workflows/release.yaml
Comment (a-khaledf) [inline .github/workflows/release.yaml:27]: environment: production and title is `Development Build`

**Review by a-khaledf — APPROVED** [review]

**Pull request merged**
